src/time_integration/methods_SSP.jl

```diff
@@ -0,0 +1,188 @@
+  du::uType
+  u_tmp::uType
+  t::RealT
+  dt::RealT # current time step
+  dtcache::RealT # ignored
+  iter::Int # current number of time steps (iteration)
+  p::Params # will be the semidiscretization from Trixi
+  sol::Sol # faked
+  alg::Alg
+  opts::SimpleIntegratorSSPOptions
+  finalstep::Bool # added for convenience
+end
+
+# Forward integrator.destats.naccept to integrator.iter (see GitHub PR#771)
+function Base.getproperty(integrator::SimpleIntegratorSSP, field::Symbol)
+  if field === :destats
+    return (naccept = getfield(integrator, :iter),)
+  end
+  # general fallback
+  return getfield(integrator, field)
+end
+
+function create_cache(cache, equations, dg, uEltype)
+  A4dp1_x = Array{uEltype, 4}
+  A4dp1_y = Array{uEltype, 4}
+  flux_antidiffusive1_threaded = A4dp1_x[A4dp1_x(undef, nvariables(equations), nnodes(dg)+1, nnodes(dg), nelements(dg, cache)) for _ in 1:Threads.nthreads()]
+  flux_antidiffusive2_threaded = A4dp1_y[A4dp1_y(undef, nvariables(equations), nnodes(dg), nnodes(dg)+1, nelements(dg, cache)) for _ in 1:Threads.nthreads()]
+
+  return (; flux_antidiffusive1_threaded, flux_antidiffusive2_threaded)
+end
+
+"""
+    solve_IDP()
+
+The following structures and methods provide a implementation of the
+SSP Runge-Kutta method of
+
+    Shu, Osher (1988) Efficient Implementation of Essentially Non-oscillatory Shock-Capturing Schemes, eq. 2.18.
+"""
+function solve_IDP(ode::ODEProblem, semi; dt, callback=nothing, kwargs...)
+  # TODO: Maybe add alg to dependency
+  alg = SSPRK33_ShuOsher()
+
+  u = copy(ode.u0)
+  du = similar(u)
+  u_tmp = similar(u)
+  t = first(ode.tspan)
+  iter = 0
+  integrator = SimpleIntegratorSSP(u, du, u_tmp, t, dt, zero(dt), iter, ode.p,
+                  (prob=ode,), alg,
+                  SimpleIntegratorSSPOptions(callback, ode.tspan; kwargs...), false)
+
+  if callback isa CallbackSet
+    for cb in callback.continuous_callbacks
+      error("unsupported")
+    end
+    for cb in callback.discrete_callbacks
+      cb.initialize(cb, integrator.u, integrator.t, integrator)
+    end
+  elseif !isnothing(callback)
+    error("unsupported")
+  end
+
+  solve!(integrator)
+end
+
+function solve!(integrator::SimpleIntegratorSSP)
+  @unpack prob = integrator.sol
+  @unpack alg = integrator
+  t_end = last(prob.tspan)
+  callbacks = integrator.opts.callback
+
+  integrator.finalstep = false
+  @trixi_timeit timer() "main loop" while !integrator.finalstep
+    if isnan(integrator.dt)
+      error("time step size `dt` is NaN")
+    end
+
+    # if the next iteration would push the simulation beyond the end time, set dt accordingly
+    if integrator.t + integrator.dt > t_end || isapprox(integrator.t + integrator.dt, t_end)
+      integrator.dt = t_end - integrator.t
+      terminate!(integrator)
+    end
+
+    @trixi_timeit timer() "RK stage" begin
+    prob.f(integrator.du, integrator.u, integrator.p, integrator.t)
+    integrator.u_tmp = integrator.u + integrator.dt * integrator.du
+    end
```

Comment:
```suggestion
    @trixi_timeit timer() "RK stage" begin
      prob.f(integrator.du, integrator.u, integrator.p, integrator.t)
      @. integrator.u_tmp = integrator.u + integrator.dt * integrator.du
    end
```
Does this even work efficiently or does it allocate a huge amount of memory as it is?

Reply:
It actually wasn't that bad before. But now it's better, thanks.